### PR TITLE
Add optional support for structured logging from Tasks.

### PR DESCRIPTION
**Pull request by dlorenc**
# Changes

This commit adds a new configmap that contains a setting that can switch the
logging mode of our entrypoint runner to "structured". If this mode is set, the
runners will read lines from stdout/stderr and wrap them in a json object before
logging.

The json object contains a 'message' key, which contains the actual log line,
and then key/value pairs for each tekton.dev/ label is attached to the Taskrun.

See this document for more context:
https://docs.google.com/document/d/19f2PpdxlzOKSKpCDAvZnQg2aCNjtLpFjrFxMiumllcE/edit?ts=5dcf1c5a

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
There is a new config map (config-entrypoint) to hold configuration for the Tekton entrypoint system. There is currently one setting there (runner: structured|default) that controls which logging output mode is used. If the structred setting is used, Tekton will convert raw stdout lines to json objects.
```

### CODE REVIEW

File: cmd/entrypoint/main.go [hunk 1]
```diff
@@ -33,26 +32,35 @@ var (
...
 func main() {
 	flag.Parse()
 
+	var r entrypoint.Runner
+	switch *runner {
+	case "logging":
+		r = &entrypoint.LoggingRunner{}
+	case "default":
+		fallthrough
+	default:
```
Comment (ImJasonH) [inline cmd/entrypoint/main.go]: Should this fail instead?

File: pkg/reconciler/taskrun/entrypoint/entrypoint.go [hunk 2]
```diff
@@ -141,12 +147,13 @@ func RedirectStep(cache *Cache, stepNum int, step *v1alpha1.Step, kubeclient kub
 	if stepNum == 0 {
 		step.VolumeMounts = append(step.VolumeMounts, downwardMount)
 	}
+	step.VolumeMounts = append(step.VolumeMounts, logCfgMount)
```
Comment (ImJasonH) [inline pkg/reconciler/taskrun/entrypoint/entrypoint.go]: Can we avoid mounting this if we're not using the logging entrypoint runner? Do we care?

File: pkg/reconciler/taskrun/entrypoint/entrypoint.go [hunk 3]
```diff
@@ -41,6 +41,8 @@ const (
 	DownwardMountName      = "downward"
 	DownwardMountPoint     = "/builder/downward"
 	DownwardMountReadyFile = "ready"
+	LogCfgMount            = "log-config"
```
Comment (ImJasonH) [inline pkg/reconciler/taskrun/entrypoint/entrypoint.go]: This code has all moved to pkg/pod, but it looks like this change should be just as easy to make there.

File: cmd/entrypoint/main.go (hunk 1 above)
Comment (sbwsg) [reply cmd/entrypoint/main.go]: Looks like @dlorenc added the comment above this case to explain its behaviour.  @ImJasonH are you happy with this to go in as-is?

Apart from this question I'm ready to lgtm.

Comment (bobcatfish) [reply cmd/entrypoint/main.go]: im still not clear on this but dont really want to leave this open indefinitely - i do feel like not failing when the logging isnt configured correctly probably makes sense (if that is whats happening here?

**Comment by bobcatfish** [comment]
welcome back @dlorenc :D

File: pkg/pod/pod.go [hunk 4]
```diff
@@ -77,6 +79,14 @@ var (
 	}}
 )
 
+func getRunner(k kubernetes.Interface) string {
+	cfg, err := configmap.Load("config-entrypoint")
+	if err != nil {
+		return ""
+	}
+	return cfg["runner"]
+}
```
Comment (bobcatfish) [inline pkg/pod/pod.go:88]: this seems a bit different from how we interact with config maps in other parts of the code, i think for other config maps we use https://github.com/tektoncd/pipeline/blob/master/pkg/apis/config/default.go

File: pkg/entrypoint/runner.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	for s.Scan() {
+		labels["message"] = s.Text()
+		b, err = json.Marshal(labels)
+		if err != nil {
+			log.Printf("error marshalling text %v: %s", labels, err)
+			return
+		}
+		fmt.Fprintln(w, string(b))
+	}
+}
```
Comment (bobcatfish) [inline pkg/entrypoint/runner.go:91]: can we have unit tests for this file?

**Comment by bobcatfish** [comment]
@dlorenc are you planning to get back to this or would it make sense to close it for now?

**Comment by sbwsg** [comment]
Closing for now but please reopen to take it over the finish line!

**Pull request closed without merging**